### PR TITLE
Glayout DRC and LVS in CI

### PR DESCRIPTION
**Pull request by chetanyagoyal**
Implements the following changes - 

1. Create a pythonic parser file that: 
- Writes the specified component's gds
- Writes its netlist
- Validates the DRC results by looking at the number of lines in the reports and the magic subprocess return codes
- Validates the LVs results by parsing the report for exceptions and checks the magic (extraction) and netgen (LVS) subprocess return codes
2. Create a `yaml` workflow file to: 
- pull the stable docker image
- run the parser file by specifying the component to run drc and lvs for
3. makes the antenna diodes optional in diff_pair_ibias_netlist to allow for NoneType being passed (when `with_antenna_diodes=0` in opamp)
4. updates the `diff_pair_ibias` code to account for antenna diodes being passed as 0
5. Update the differential to single ended converter netlist 
- `.subckt` definition now matches the parameters being passed
- this makes things the `DIFF_TO_SIGLE` netlist LVS clean
6. Switch around pin around in mimcap netlist reference in opamp gain stage
7. Update the current mirror netlist:
- to use length as minimum transistor length if passed as None
- pass length and width to the `.subckt` statements 
8. General quality of life improvements in lvs, drc (magic and klayout) in `mappedpdk`
9. modify fet netlist:
- to pass length, width, multipliers and fingers to `.subckt` calls
  - length is ltop
  - width is mtop
  - fingers * multipliers is mtop (the number of parallel devices)
  - multipliers is dummy multipliers
- divide the number of main fet multipliers by two to fix the reported width from cdl always being double of whats in the extracted spice
10. add `with_pins` argument in `generate_netlist` to allow removal of top level's pins

### CODE REVIEW

**Comment by msaligane** [comment]
is this ready for review?

**Comment by d-m-bailey** [comment]
@msaligane the `mappedpdk.py` file contains the following extraction commands
```
select top cell
extract all
ext2spice lvs
ext2spice -o {str(lvsmag_path)}

load {design_name}
extract all
ext2spice lvs 
ext2spice rthresh 0
ext2spice cthresh 0
ext2spice -o {str(pex_path)}

load {design_name}
extract all
ext2spice cthresh 0
ext2spice -o {str(sim_path)}
```
I think the same extraction is happening 3 times. I don't think the ext2spice modify the extracted data (`*.ext` files), just how that information is netlisted.

The default extract option is to connect virtual nets (nets with the same name) that are not physically connected.

You might want to use the [detailed parasitic resistance flow described here](http://opencircuitdesign.com/magic/commandref/extresist.html).

The netgen lvs command used is a one line command that should work for simple circuits. Multiple netlists and mixed signal designs require extra handling.

You might want to add a check for unexpected blackboxes. netgen will sometimes match an empty subckt to an actual layout or visa versa because it assumes that one or the other is a blackbox.

**Comment by chetanyagoyal** [comment]
> Contributor

@d-m-bailey I addressed the repeated `.ext` extractions in [this commit](https://github.com/idea-fasoc/OpenFASOC/pull/304/commits/a701be0198b1cd06bea2be3b8309771590140dad). 

- Adding `ext2resist` didn't seem to make a difference in the report. 
- From what I read in the documentation, adding `-blackbox` seemed like a good check to add

Thank you for the help!

**Comment by chetanyagoyal** [comment]
@msaligane this is ready to merge

**Comment by alibillalhammoud** [comment]
@srpathen @arldai can you take a look?

File: .github/scripts/expected_sim_outputs/opamp/means.json
Comment (alibillalhammoud) [inline .github/scripts/expected_sim_outputs/opamp/means.json:1]: remove or explain why this is here

File: .github/scripts/expected_sim_outputs/opamp/variances.json
Comment (alibillalhammoud) [inline .github/scripts/expected_sim_outputs/opamp/variances.json:1]: remove or explain why this is here

File: .github/scripts/test_glayout.py
Comment (alibillalhammoud) [inline .github/scripts/test_glayout.py:1]: The test PR should be merged before this PR, so I will ignore this file for now

**Review by alibillalhammoud — CHANGES_REQUESTED** [review]
Major changes are fine, I think we should hold off on merging this PR until the other PRs are merged.

File: .github/scripts/expected_sim_outputs/opamp/means.json
Comment (chetanyagoyal) [reply .github/scripts/expected_sim_outputs/opamp/means.json:1]: this is used for the opamp spice testbench simulations as a lookup for means/variances of the 100 runs i did

File: .github/scripts/expected_sim_outputs/opamp/variances.json
Comment (chetanyagoyal) [reply .github/scripts/expected_sim_outputs/opamp/variances.json:1]: same here

**Pull request closed without merging**
